core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py

```diff
@@ -39,7 +39,8 @@ def __init__(self, model_base: AbstractModel, random_state=0, **kwargs):
         self._k_per_n_repeat = []  # k-fold used for each n_repeat. == [5, 10, 3] if first kfold was 5, second was 10, and third was 3
         self._random_state = random_state
         self.low_memory = True
-        self.bagged_mode = None
+        self._bagged_mode = None
+        self._child_oof = False  # Whether the OOF preds were taken from a single child model (Assumes child can produce OOF preds without bagging).
```

Comment:
Should there eventually be a separate class for models that can produce OOF predictions without bagging? (RF/XT could eventually also do this in addition to KNN). Seems like these could just be of the same class as our regular models (single-copy, not bagged), adding them under BaggedEnsemble might be misleading since these are neither  ensembles nor using bagging?

Reply:
Alternatively could make BaggedEnsemble a subclass of some shared parent-class (eg. OOFModel) that's capable of producing held-out prediction for any requested datapoint?

Reply:
These are all reasonable, just will take quite a bit of thought and coding to make sure its done right.

Reply:
The main thing is we want to keep the option available to bag these models that could also get OOF, for the purposes of consistency and comparisons. Furthermore, we should have the option to get training preds from models that don't support LOO and just have the predictions be overfit. For large enough datasets, it might be the case that we can get away with this if the training vs val scores are not much different.

For these reasons, I don't think it's the right idea to subclass the models themselves. A tag + implementation of the get_oof method should suffice. I think `OOFModel` may be the direction to go, such that `BaggedEnsemble` inherits from it. This can also mean that all `_FULL` models instead become `OOFModel` instead of `BaggedEnsemble`.

Reply:
The difficulty from this is that StackerEnsembleModel inherits from BaggedEnsemble, but in fact these single models can also be stackers. Unless we have multiple StackerEnsembleModels, which is ugly, we would have to re-think how our stacker models operate.

Reply:
One option is to delete stacker ensemble model and instead move all stacking logic to trainer, therefore completely hiding stacking logic from models.

Reply:
I agree with "tag + implementation of the get_oof method should suffice".  Perhaps with the addition of OOFModel, StackerEnsembleModel could simply inherit from OOFModel instead of BaggedEnsemble? 

Understand it'll take significant code to add OOFModel, so may be only worthwhile once we have another model like RF/XT/LR also not requiring bagging to get OOF.  At least for this PR, maybe just add comment in BaggedEnsembleModel that describes the special exception for using single child model for all OOF-predictions like in the case of KNeighbors.

Reply:
The problem with "StackerEnsembleModel could simply inherit from OOFModel instead of BaggedEnsemble" is that then BaggedEnsembleModel wouldn't be part of StackerEnsembleModel.

Reply:
Right it may be annoying to change StackerEnsemble too much.
Another option could be to simply change the name of the class `BaggedEnsemble` to be broader like `OOFModel`, with some attribute for distinguishing whether it's based on bagging or not.

Or could keep BaggedEnsemble and adapt Stacker to work with both BaggedEnsemble base models as well as regular base models that offer  `predict_oof()` method (where these are not subclass of BaggedEnsemble). But heterogeneity of stacking BaggedEnsemble and individual models together may be annoying as well...

Reply:
Added comment about child_oof, added tags. Will leave `OOFModel` to a separate PR as it only becomes important if we have widespread usage of custom `OOF` generation strategies.

Reply:
Sounds good, I'd also edit the comment at the top of BaggedEnsemble class, as the current description is no longer accurate.

Reply:
Edited the doc